add withEncode function

EricCrosson opened the pull request:
`withEncode` is similar to `withValidate`:

`withEncode` clones an existing codec and overrides the existing
`encode` function with the given `encode` function. This can be useful
when you want to use `t.type` to combinatorially create codecs but
want to leverage the `encode` function to encode into a different type.

gcanti commented:
Thanks @EricCrosson, I would add an optional `name` argument though: 

```ts
export function withEncode<A, O, I, P>(
  codec: t.Type<A, O, I>,
  encode: (a: A) => P,
  name: string = codec.name
): t.Type<A, P, I> {
  return new t.Type(name, codec.is, codec.validate, encode)
}
```

EricCrosson committed:
feat: add withEncode function

`withEncode` is similar to `withValidate`:

`withEncode` clones an existing codec and overrides the existing
`encode` function with the given `encode` function. This can be useful
when you want to use `t.type` to combinatorially create codecs but
want to leverage the `encode` function to encode into a different type.

EricCrosson commented:
Excellent! That's a much better implementation too. Thanks for the suggestion @gcanti , I've pushed these updates to the PR
